cmd/outline-ss-server/main.go

```diff
@@ -59,38 +63,28 @@ func init() {
 	logger = logging.MustGetLogger("")
 }
 
-type ssPort struct {
-	tcpListener *net.TCPListener
-	packetConn  net.PacketConn
-	cipherList  service.CipherList
+type ssListener struct {
+	io.Closer
+	cipherList service.CipherList
 }
 
 type SSServer struct {
 	natTimeout  time.Duration
 	m           *outlineMetrics
 	replayCache service.ReplayCache
-	ports       map[int]*ssPort
+	listeners   map[string]*ssListener
 }
 
-func (s *SSServer) startPort(portNum int) error {
-	listener, err := net.ListenTCP("tcp", &net.TCPAddr{Port: portNum})
+func (s *SSServer) startDirectTCP(addr *net.TCPAddr, cipherList service.CipherList) (*net.TCPListener, error) {
```

Comment:
I'm not sure this will scale. Are you going to have a method for each listener type?

Perhaps `serve(listener, cipherList)`, and have the logic switch based on the listener type (packet or stream)?